File: backend/app/recommend.py
from pydantic import BaseModel
from typing import List
from app.database import supabase
import time


class UserInput(BaseModel):
    categories: List[str]
    monthly_spend: int
    priority: str
    fee_tolerance: str
    income: str


def generate_recommendation(data: UserInput):

    start = time.time()

    category = data.categories[0]

    result = (
        supabase.table("cards")
        .select("*")
        .execute()
    )

    cards = result.data

    if not cards:
        return {
            "recommended_card": "No Match Found",
            "reason": "No cards available",
            "confidence": 0
        }

    income_map = {
        "lt6": 500000,
        "6to12": 900000,
        "12to25": 1800000,
        "gt25": 3000000
    }

    income = income_map.get(data.income, 500000)

    scored_cards = []

    for card in cards:

        score = 0

        # Income eligibility
        if income >= card["min_income"]:
            score += 40

        # Dynamic reward column
        reward_col = f"{category}_reward"
        reward = float(card.get(reward_col, 0) or 0)

        score += int(reward * 5)

        # Lounge bonus
        if card["lounge_access"]:
            score += 15

        # Fee preference
        if data.fee_tolerance == "low":
            if card["annual_fee"] <= 1000:
                score += 20

        elif data.fee_tolerance == "medium":
            if card["annual_fee"] <= 3000:
                score += 20

        else:
            score += 20

        scored_cards.append({
            "card": card,
            "score": score
        })

    scored_cards.sort(
        key=lambda x: x["score"],
        reverse=True
    )

    top = scored_cards[0]

    alts = []

    for item in scored_cards[1:3]:

        alt_card = item["card"]

        reward_col = f"{category}_reward"
        alt_reward = float(
            alt_card.get(reward_col, 0) or 0
        )

        alts.append({
            "card": {
                "id": str(alt_card["id"]),
                "name": alt_card["card_name"],
                "annualFee": alt_card["annual_fee"],
                "joiningFee": alt_card["joining_fee"],
                "feeWaiverSpend": alt_card["fee_waiver_spend"],
                "minIncome": alt_card["min_income"],
                "forexMarkup": float(alt_card["forex_markup"]),
                "topPerks": alt_card["top_perks"] or [],
                "accent": alt_card["accent"],
                "rewardCurrency": alt_card["reward_currency"],
                "loungeDomestic": alt_card["lounge_domestic"],
                "loungeIntl": alt_card["lounge_intl"]
            },
            "score": item["score"],
            "confidence": item["score"],
            "reasons": [
                f"{alt_reward}% rewards on {category} spends",
                f"₹{alt_card['annual_fee']} annual fee",
                f"{alt_card['lounge_domestic']} lounge visits included"
            ],
            "estAnnualReward": int(
                data.monthly_spend * 12 * (alt_reward / 100)
            )
        })

    card = top["card"]

    reward_col = f"{category}_reward"

    top_reward = float(
        card.get(reward_col, 0) or 0
    )

    insert_result = (
        supabase.table("recommendations")
        .insert({
            "card_id": card["id"],
            "score": top["score"]
        })
        .execute()
    )

    print("TOP CARD =", card["card_name"])
    print("CATEGORY =", category)
    print("TOP REWARD =", top_reward)
    print("INSERT RESULT =", insert_result)
    print("TIME TAKEN =", time.time() - start)

    return {
        "top": {
            "card": {
                "id": str(card["id"]),
                "name": card["card_name"],
                "annualFee": card["annual_fee"],
                "joiningFee": card["joining_fee"],
                "feeWaiverSpend": card["fee_waiver_spend"],
                "minIncome": card["min_income"],
                "forexMarkup": float(card["forex_markup"]),
                "topPerks": card["top_perks"] or [],
                "accent": card["accent"],
                "rewardCurrency": card["reward_currency"],
                "loungeDomestic": card["lounge_domestic"],
                "loungeIntl": card["lounge_intl"]
            },
            "score": top["score"],
            "confidence": top["score"],
            "reasons": [
                f"{top_reward}% rewards on {category} spends",
                f"₹{card['annual_fee']} annual fee",
                f"{card['lounge_domestic']} lounge visits included"
            ],
            "estAnnualReward": int(
                data.monthly_spend * 12 * (top_reward / 100)
            )
        },
        "alts": alts
    }

File: frontend/src/lib/recommend.ts
import { CARDS, type Card, type SpendCategory } from "@/data/cards";
import type { Answers } from "@/store/useFlowStore";

export interface Recommendation {
  card: Card;
  score: number;
  confidence: number;
  reasons: string[];
  estAnnualReward: number;
}

const INCOME_MAP: Record<string, number> = {
  lt6: 500000,
  "6to12": 900000,
  "12to25": 1800000,
  gt25: 3000000,
};

const CATEGORY_LABEL: Record<SpendCategory, string> = {
  travel: "travel",
  food: "dining",
  shopping: "online shopping",
  fuel: "fuel",
  bills: "utility bills",
  groceries: "groceries",
};

function scoreCard(card: Card, a: Answers) {
  const income = a.income ? INCOME_MAP[a.income] : 600000;
  // Hard eligibility (soft penalty rather than exclude — prototype)
  const eligible = income >= card.minIncome * 0.7;
  const overQualified = income >= card.minIncome * 2.5;

  // Category fit (40) — weighted by stated monthly spend distribution
  const cats = a.categories.length ? a.categories : (["shopping", "food"] as SpendCategory[]);
  const catScore =
    (cats.reduce((sum, c) => sum + (card.rewardsRate[c] ?? 0), 0) / (cats.length * 10)) * 40;

  // Lifestyle (25)
  const lifeScore = a.priority && card.lifestyle.includes(a.priority) ? 25 : 6;

  // Fee tolerance (15) — sharper differentiation
  let feeScore = 0;
  if (a.feeTolerance === "free") feeScore = card.annualFee <= 750 ? 15 : card.annualFee <= 2000 ? 6 : 0;
  else if (a.feeTolerance === "under5k") feeScore = card.annualFee <= 5000 ? 15 : card.annualFee <= 12500 ? 7 : 2;
  else feeScore = card.annualFee >= 5000 ? 15 : 10; // worth it — favors premium

  // Redemption fit (10)
  const redeem =
    a.priority === "cashback"
      ? card.rewardCurrency === "Cashback" ? 10 : 2
      : a.priority === "miles"
      ? card.rewardCurrency.toLowerCase().includes("miles") ? 10 : 3
      : 7;

  // Eligibility (10) — penalize over-qualification mildly to avoid recommending entry cards to HNI
  const elig = eligible ? (overQualified && card.annualFee < 1000 ? 4 : 10) : 1;

  const total = catScore + lifeScore + feeScore + redeem + elig;
  return { total, eligible };
}

function buildReasons(card: Card, a: Answers): string[] {
  const out: string[] = [];
  const topCat = a.categories
    .map((c) => ({ c, r: card.rewardsRate[c] ?? 0 }))
    .sort((x, y) => y.r - x.r)[0];
  if (topCat && topCat.r > 0) {
    out.push(
      `You spend big on ${CATEGORY_LABEL[topCat.c]} — this card returns ${topCat.r}% there, the highest among the cards we matched.`,
    );
  } else {
    out.push(`Strong all-round rewards on the categories you spend on most.`);
  }
  if (a.priority === "lounge" || a.priority === "status" || a.priority === "miles") {
    if (card.loungeIntl >= 4 || card.loungeIntl >= 99) {
      out.push(`Unlimited domestic + ${card.loungeIntl >= 99 ? "unlimited" : card.loungeIntl} international lounge visits — matches your travel-first vibe.`);
    } else if (card.loungeDomestic > 0) {
      out.push(`${card.loungeDomestic >= 99 ? "Unlimited" : card.loungeDomestic} domestic lounge visits a year.`);
    }
  }
  if (a.feeTolerance === "free" && card.annualFee <= 750) {
    out.push(`₹${card.annualFee} annual fee — within your "fee-light" preference.`);
  } else if (card.feeWaiverSpend) {
    out.push(`The ₹${card.annualFee.toLocaleString("en-IN")} fee is waived on ₹${(card.feeWaiverSpend / 100000).toFixed(1)}L annual spend — easily clearable at your level.`);
  } else {
    out.push(`Welcome benefits typically recover the joining fee within your first 2 months.`);
  }
  return out.slice(0, 3);
}

function estimateAnnualReward(card: Card, a: Answers) {
  const monthly = a.monthlySpend;
  const cats = a.categories.length ? a.categories : (["shopping", "food"] as SpendCategory[]);
  const avgRate = cats.reduce((s, c) => s + (card.rewardsRate[c] ?? 0), 0) / cats.length;
  return Math.round(monthly * 12 * (avgRate / 100));
}

export function recommend(a: Answers): { top: Recommendation; alts: Recommendation[] } {
  const ranked = CARDS.filter((c) => !a.existingCards.includes(c.id))
    .map((card) => {
      const { total, eligible } = scoreCard(card, a);
      return {
        card,
        score: total,
        confidence: Math.min(98, Math.round(total + (eligible ? 6 : 0))),
        reasons: buildReasons(card, a),
        estAnnualReward: estimateAnnualReward(card, a),
      };
    })
    .sort((a, b) => b.score - a.score);

  return { top: ranked[0], alts: ranked.slice(1, 3) };
}
